linkerd/app/core/src/proxy/server.rs

```diff
@@ -192,35 +192,48 @@ where
         Box::pin(async move {
             match http_version {
                 HttpVersion::Http1 => {
-                    // Enable support for HTTP upgrades (CONNECT and websockets).
-                    let svc = upgrade::Service::new(http_svc, drain.clone());
-                    let conn = builder
-                        .http1_only(true)
-                        .serve_connection(io, HyperServerSvc::new(svc))
-                        .with_upgrades();
-
-                    Ok(Box::pin(async move {
-                        drain
-                            .watch(conn, |conn| Pin::new(conn).graceful_shutdown())
-                            .instrument(info_span!("h1"))
-                            .await?;
+                    let serve = async move {
+                        // Enable support for HTTP upgrades (CONNECT and websockets).
+                        let svc = upgrade::Service::new(http_svc, drain.clone());
+                        let mut conn = builder
+                            .http1_only(true)
+                            .serve_connection(io, HyperServerSvc::new(svc))
+                            .with_upgrades();
+
+                        tokio::select! {
+                            res = &mut conn => {
+                                res?
+                            }
+                            () = drain.into_future() => {
+                                Pin::new(&mut conn).graceful_shutdown()
+                            }
```

Comment:
I don't think this has the same behavior as the way we were using `Drain::watch`  previously. `watch` not only triggers the graceful shutdown but also causes the drained future to wait until the connection's graceful shutdown has _completed_. On the other hand, I think this just triggers the graceful shutdown without counting it against the drain's completion.

I believe that this may result in the proxy exiting before it has performed a graceful shutdown on all connections,